### ISSUE TITLE
The render process take much CPU time and very slow

### ISSUE THREAD

**Issue by testforstephen** [microsoft/vscode#117327]
- VS Code Version: 1.53.2
- OS Version: macOS Catalina 10.15.5

I'm working on a Java project https://github.com/eclipse/eclipse.jdt.ls, which includes about 6K+ Java files. It works well in 1.52. But after upgrading VS Code to 1.53, editing at the same workspace becomes pretty slow. Especially after i edit some Java files for a while, the whole VS Code starts to become pretty slow.

And i saw the render process took much CPU resources.
<img width="859" alt="Screen Shot 2021-02-23 at 15 03 45" src="https://user-images.githubusercontent.com/14052197/108812411-09137800-75ea-11eb-8198-f8a6d1b4c208.png">

Below is the CPU profile result by [profiling the render process](https://github.com/microsoft/vscode/wiki/Performance-Issues#profiling-the-renderer-process).

[CPU-20210223T150659.cpuprofile.txt](https://github.com/microsoft/vscode/files/6027260/CPU-20210223T150659.cpuprofile.txt)

**Comment by jrieken** [comment]
It looks like the extension host is send a lot of data into the renderer. @testforstephen do you use an extension that contributes editor decorations, like rainbow brackets or git lens' inline decorations or anything like that?

<img width="973" alt="Screenshot 2021-02-23 at 17 05 32" src="https://user-images.githubusercontent.com/1794099/108871342-58ad7000-75f9-11eb-99fa-28baa9aa7d93.png">

**Comment by jrieken** [comment]
@testforstephen Can you run vscode with this flag `--no-cached-data` profile again. That should yield better results. And look out for fishy extensions and would be causing this

**Comment by testforstephen** [comment]
I installed GitLens extension, which will show inline message inside editor. Let me disable it and try again. Will update here if i get more information.

**Comment by andxu** [comment]
[CPU-20210303T111853.cpuprofile.txt](https://github.com/microsoft/vscode/files/6073405/CPU-20210303T111853.cpuprofile.txt)
I encounter this problem, here is the cpu profile

**Comment by CsCherrYY** [comment]
I have the same problem as well. Profile is attached.
[CPU-20210303T112914.cpuprofile.txt](https://github.com/microsoft/vscode/files/6073428/CPU-20210303T112914.cpuprofile.txt)

**Comment by jrieken** [comment]
Thanks. These profiles are much more help, tho I can't put my finger onto something yet. The problem is that it takes a very long time to receive data from the extension (which usually means it too much data) but I don't know yet what kind of data that is. A few questions

* does this only happen with a certain language, like Java
* do you have a very large amount completions, like more than 100000 items?
* does this reproduce when you disable semantic highlights? Please try with `editor.semanticTokenColorCustomizations: {enabled: false}`

**Comment by CsCherrYY** [comment]
@jrieken 
Thanks a lot for the reply. Here are some findings about your questions:
> - does this only happen with a certain language, like Java

Yes. There is no performance issue when I was editing typescript file.
> - do you have a very large amount completions, like more than 100000 items?

Code completion in a middle amount type, which has 1456 results in total will cause about 10 seconds' freeze. Language server's log shows that the calculation time of the completion results is about 300ms, the remaining time is in VS Code.

Another finding is that when I disable the Java code completion, there is no performance issue and it works well.
> - does this reproduce when you disable semantic highlights? Please try with editor.semanticTokenColorCustomizations: {enabled: false}

I have tried to disable the semantic highlighting, but it has no effect.

**Comment by jrieken** [comment]
> Code completion in a middle amount type, which has 1456 results in total will cause about 10 seconds' freeze. Language server's log shows that the calculation time of the completion results is about 300ms, the remaining time is in VS Code.

How do you measure those 10 seconds? Is there an easy way for me to reproduce this? Can you setup a sample project that reproduces this so that I can test it locally?

**Comment by CsCherrYY** [comment]
Just estimated the time from the VS Code UI freeze to it became active again. You can follow these steps to reproduce this:
- Install the newest version of Java extension `Language Support for Java(TM) by Red Hat`
- Git clone https://github.com/spring-projects/spring-petclinic.git
- Open the project folder with VS Code
- Wait until the project is imported successfully (the thumb shows in the status bar)
![Screenshot 2021-03-03 160616](https://user-images.githubusercontent.com/45906942/109773735-810d1e00-7c3a-11eb-8974-c88e5d85b99c.jpg)
- Open `spring-petclinic/src/main/java/org/springframework/samples/petclinic/Owner.java`, in line 100, method `getPets()`, try to add a new variable. When you input some class names, like `String`, you will encounter this issue, like the following gif.
![issue](https://user-images.githubusercontent.com/45906942/109774912-cb42cf00-7c3b-11eb-9376-0821f5f1d239.gif)

**Comment by jrieken** [comment]
This is quite OK for me.. Any special setting that you are using?

https://user-images.githubusercontent.com/1794099/109776374-c9a8e500-7c02-11eb-86ad-a071978d65fa.mov

**Comment by jrieken** [comment]
@CsCherrYY One more thing: in the profile that you have attached I have seen that the extension host was blocked and automatically profiled. Can you open the renderer log (F1 > Show Logs > Window) and then search for messages that start with "UNRESPONSIVE extension host," That message will mention a CPU profile that was automatically created for the extension host. Can you also attach that file. Thanks

**Comment by CsCherrYY** [comment]
Sure. This is your mentioned file:
[exthost-6c71f9.cpuprofile.txt](https://github.com/microsoft/vscode/files/6074699/exthost-6c71f9.cpuprofile.txt)

We find that the root cause maybe the setting `java.trace.server`. Could you please try to change this setting to `verbose` to see if this issue can be reproduced or not?

**Comment by jrieken** [comment]
Are you using `console.log` and friends for logging these messages?

**Comment by jrieken** [comment]
I can reproduce. This is due to very intense usage of `console`. You can (and should) fix this in your extension by using the output channel API only, and not use console (which is recommended practise anyways). 

<img width="434" alt="Screenshot 2021-03-03 at 10 21 38" src="https://user-images.githubusercontent.com/1794099/109783399-3ffd1580-7c0a-11eb-8685-367a66250a3c.png">

Assigning to @alexdima for protecting us from this

**Comment by testforstephen** [comment]
@jrieken Great! Thanks to help us to find out the culprit.

Yes, the Java language server will print the LSP verbose log to both output channel and the console of Developer Tools. So looks like it's necessary to not print it to console.

**Comment by jrieken** [comment]
> So looks like it's necessary to not print it to console.

Yeah, actually for two reasons: Performance. And that's true across the board, also Chrome Dev Tools are dying with this amount of data. The other reason is that the console is "shared resource" there also other extensions and us and it would be a huge mess if everyone does their verbose logging here.
